central/complianceoperator/v2/profiles/service/service_impl.go

```diff
@@ -96,13 +126,50 @@ func (s *serviceImpl) ListProfileSummaries(ctx context.Context, request *v2.Clus
 		return nil, errors.Wrap(errox.InvalidArgs, "cluster is required")
 	}
 
-	profiles, err := s.complianceProfilesDS.GetProfilesByClusters(ctx, request.GetClusterIds())
+	// Fill in Query.
+	parsedQuery, err := search.ParseQuery(request.GetQuery().GetQuery(), search.MatchAllIfEmpty())
+	if err != nil {
+		return nil, errors.Wrapf(errox.InvalidArgs, "Unable to parse query %v", err)
+	}
+
+	// To get total count, need the parsed query without the paging.
+	countQuery := parsedQuery.Clone()
+
+	// Fill in pagination.
+	paginated.FillPaginationV2(parsedQuery, request.GetQuery().GetPagination(), maxPaginationLimit)
+	// make sure we sort by profile name at a minimum
+	if parsedQuery.GetPagination().GetSortOptions() == nil {
+		parsedQuery.Pagination.SortOptions = []*v1.QuerySortOption{
+			{
+				Field: search.ComplianceOperatorProfileName.String(),
+			},
+		}
+	}
+
+	profileNames, err := s.complianceProfilesDS.GetProfilesNames(ctx, parsedQuery, request.GetClusterIds())
+	if err != nil {
+		return nil, errors.Wrapf(errox.InvalidArgs, "Unable to retrieve compliance profiles for %v", request)
+	}
+
+	// Build query to get the filtered list by profile names
+	profileQuery := search.NewQueryBuilder().AddSelectFields().AddExactMatches(search.ComplianceOperatorProfileName, profileNames...).ProtoQuery()
+	// Bring the sort options only, paging is handled in step one when we get the distinct profiles.
```

Comment:
Why handle it as a second step? Because a paginated result can not be sorted correctly?

Reply:
Since the profiles are denormalized, getting all the "distinct' data in one shot with our framework was quite cumbersome.  So the simplest path was to use 2 passes.  One to use paging and sorting to get the distinct profile names.  Then go back to get the details for just those profiles and build the response.  I reserve the right to figure out a better way later, but this is the simplest path using our current framework while supporting pagination.